### DISCUSSION TITLE
Generation of white noise

### DISCUSSION THREAD

**Discussion by PiezaKake** [bonsai-rx/bonsai#1365] [Q&A] [answered]
Hi,

I want to generate white noise in the range 2-91 kHz. The signal will be then be fed to an AnalogOutput node. What would be the best way to generate the noise?

**Comment by bruno-f-cruz** [comment, marked as answer]
Hey,
Generating a waveform in python and uploading it to bonsai using a matrixreader is likely the most customizable way. Otherwise, you can probably do something close in bonsai by sampling N samples from a normal distribution and applying a bandpass filter.

**Reply by PiezaKake** [reply]
Great, thanks!

**Comment by jbscherrer** [comment]
Depending on your experiment, it may be important to use "frozen noise", i.e. the same noise waveform each time the noise is played. In that case the pregenerated noise source that Bruno mentioned is probably the easiest.

I'm currently triggering noise bursts with 1 ms latency in my experiments, and to get that low latency I've just been driving a digital out line that triggers an external noise generator. In that case you can program an arduino with a ring buffer that will generate pseudorandom output pretty easily, or use a scientific signal generator with a noise output.

**Comment by PiezaKake** [comment]
I have generated white noise and saved in a .txt file (attached). I am trying to read it with MatrixReader at a sample rate of 200 kHz, however the values are not read correctly. I have played with the properties of the node, but I cannot get it right.
[white noise3.zip](https://github.com/bonsai-rx/bonsai/files/11533957/white.noise3.zip)

**Reply by bruno-f-cruz** [reply]
I think you generated your waveform as a text file instead of a flat binary file. If you want to use this format you should use `csvreader` operator:

![image](https://github.com/bonsai-rx/bonsai/assets/7049351/8db83fd8-6164-42b6-b5ed-33a8d23fe924)

**Reply by PiezaKake** [reply]
Can you please upload the code?

**Reply by bruno-f-cruz** [reply]
<details>
    <summary>Workflow (Copy+Paste me)</summary>
    
```xml




<?xml version="1.0" encoding="utf-8"?>
<WorkflowBuilder Version="2.7.3"
                 xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                 xmlns:io="clr-namespace:Bonsai.IO;assembly=Bonsai.System"
                 xmlns:rx="clr-namespace:Bonsai.Reactive;assembly=Bonsai.Core"
                 xmlns:dsp="clr-namespace:Bonsai.Dsp;assembly=Bonsai.Dsp"
                 xmlns="https://bonsai-rx.org/2018/workflow">
  <Workflow>
    <Nodes>
      <Expression xsi:type="io:CsvReader">
        <io:FileName />
        <io:ScanPattern>%f</io:ScanPattern>
        <io:SkipRows>0</io:SkipRows>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="rx:ToArray" />
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="dsp:ConvertFromArray">
          <dsp:Size>
            <dsp:Width>0</dsp:Width>
            <dsp:Height>0</dsp:Height>
          </dsp:Size>
          <dsp:Depth xsi:nil="true" />
          <dsp:Channels xsi:nil="true" />
        </Combinator>
      </Expression>
    </Nodes>
    <Edges>
      <Edge From="0" To="1" Label="Source1" />
      <Edge From="1" To="2" Label="Source1" />
    </Edges>
  </Workflow>
</WorkflowBuilder>


```

</details>
